oauth2_lib/fastapi.py

```diff
@@ -212,139 +196,155 @@ async def __call__(self, request: Request, token: Union[str, None] = None) -> Un
+    """
 
-    raise HTTPException(
-        status_code=HTTPStatus.FORBIDDEN,
-        detail=f"User is not allowed to access resource: {context.get('resource')} Decision was taken with id: {did}",
-    )
-
-
-def opa_decision(
-    opa_url: str,
-    oidc_security: OIDCUser,
-    auto_error: bool = True,
-    opa_kwargs: Union[Mapping[str, str], None] = None,
-) -> Callable[[Request, OIDCUserModel, AsyncClient], Awaitable[Union[bool, None]]]:
-    async def _opa_decision(
-        request: Request,
-        user_info: OIDCUserModel = Depends(oidc_security),
-        async_request: AsyncClient = Depends(_make_async_client),
-    ) -> Union[bool, None]:
-        """Check OIDCUserModel against the OPA policy.
+    def __init__(self, opa_url: str, auto_error: bool = True, opa_kwargs: Union[Mapping[str, Any], None] = None):
+        self.opa_url = opa_url
+        self.auto_error = auto_error
+        self.opa_kwargs = opa_kwargs
 
-        This is used as a security module in Fastapi projects
-        This method will make an async call towards the Policy agent.
+    async def get_decision(self, async_client: AsyncClient, opa_input: dict) -> OPAResult:
+        logger.debug("Posting input json to Policy agent", opa_url=self.opa_url, input=opa_input)
+        try:
+            result = await async_client.post(self.opa_url, json=opa_input)
+        except (NetworkError, TypeError) as exc:
+            logger.debug("Could not get decision from policy agent", error=str(exc))
+            raise HTTPException(status_code=HTTPStatus.SERVICE_UNAVAILABLE, detail="Policy agent is unavailable")
 
-        Args:
-            request: Request object that will be used to retrieve request metadata.
-            user_info: The OIDCUserModel object that will be checked
-            async_request: The httpx client.
-        """
+        json_result = result.json()
+        logger.debug("Received decision from policy agent", decision=json_result)
+        return OPAResult.parse_obj(json_result)
+
+    def evaluate_decision(self, decision: OPAResult, **context: dict[str, Any]) -> bool:
+        did = decision.decision_id
+
+        if decision.result:
+            logger.debug("User is authorized to access the resource", decision_id=did, **context)
+            return True
+
+        logger.debug("User is not allowed to access the resource", decision_id=did, **context)
+        if not self.auto_error:
+            return False
+
+        raise HTTPException(
+            status_code=HTTPStatus.FORBIDDEN,
+            detail=f"User is not allowed to access resource: {context.get('resource')} Decision was taken with id: {did}",
+        )
+
+
+class OPAAuthorization(Authorization, OPAMixin):
+    """Applies OPA decisions to HTTP requests for authorization.
 
+    Uses OAUTH2 settings and request information to authorize actions.
+    """
+
+    async def authorize(
+        self, request: HTTPConnection, user_info: OIDCUserModel = Depends(oidc_instance.authenticate)
```

Comment:
I can't seem to override the `OIDCAuth` subclass and create a `custom_oidc_instance`, because this function refers to the default `oidc_instance`. I'm trying to override the function now with our `custom_oidc_instance` and then call the function through `super()`, but FastAPI is not digging the conflicting Depends annotations.

Have you able to do this in your own orchestrator in some way?

Reply:
This turned out to be something else.

Reply:
Seems to be working this way; snippet from our surf orchestrator

```python
# surf/security.py

class SurfAuthN(OIDCAuth):

    async def userinfo(self, async_request: AsyncClient, token: str) -> SurfOIDCUserModel:
        # (removed)
        # data = result from introspect endpoint
        return SurfOIDCUserModel(**data)


surf_oidc_instance = SurfAuthN(
    openid_url=oauth2lib_settings.OIDC_BASE_URL,
    openid_config_url=oauth2lib_settings.OIDC_CONF_URL,
    resource_server_id=oauth2lib_settings.OAUTH2_RESOURCE_SERVER_ID,
    resource_server_secret=oauth2lib_settings.OAUTH2_RESOURCE_SERVER_SECRET,
    oidc_user_model_cls=SurfOIDCUserModel,
)

class SurfAuthZ(OPAAuthorization):

    async def authorize(
        self, request: HTTPConnection, user_info: SurfOIDCUserModel = Depends(surf_oidc_instance.authenticate)
    ) -> Optional[bool]:
        return await super().authorize(request, user_info)


surf_opa_instance = SurfAuthZ(opa_url=oauth2lib_settings.OPA_URL)
```

And in an external env variables file

```
AUTHENTICATION_INSTANCE: surf.security.surf_oidc_instance
AUTHORIZATION_INSTANCE: surf.security.surf_opa_instance
OPA_URL: http://127.0.0.1:8181/v1/data/automation/authorization/allow
```

Reply:
In orchestrator-core I had to change the api code like below;

```py
#from orchestrator.security import get_authorization
from orchestrator.security import get_authorization_function

# authorization = get_authorization()

api_router = APIRouter()
api_router.include_router(
    fixed_input.router,
    prefix="/fixed_inputs",
    tags=["Core", "Fixed Inputs"],
    #dependencies=[Depends(authorization.authorize)],
    dependencies=[Depends(get_authorization_function)],
)
```

Where `get_authorization_function()` calls the `get_authorization()` at runtime instead of import-time. Otherwise the `authorization` object contained the OPA instance from oauth2-lib instead of our custom one.

Reply:
Although I have to conduct more tests, at a glance it seems to work :)

I am wondering though if there's a way to override/extend the Auth classes without the need for a global variable and the dynamic module loading. I recognize the difficulty being how functions are passed into the FastAPI router/endpoints, but perhaps there is a way to solve that. Want to look into this together with @dmarjenburgh next week.

Reply:
I tried many ways to see if I could avoid using Depends from FastAPI. There are some challenges I faced:

- **Keeping OPA Code:** The OPA code was really good, so I wanted to keep it as much as possible without removing anything.

- **Working with MyPy:** It was difficult to make MyPy happy while using FastAPI’s Depends, but I managed to make it work.

It’s important to know that using OPA is not necessary for everyone, but I still found it useful.

I also got some ideas from how Django lets you specify the name of the module you want to use, which helped me figure out how to manage dependencies by specifying the FQDN in GSO.

Reply:
> In orchestrator-core I had to change the api code like below;
> 
> ```python
> #from orchestrator.security import get_authorization
> from orchestrator.security import get_authorization_function
> 
> # authorization = get_authorization()
> 
> api_router = APIRouter()
> api_router.include_router(
>     fixed_input.router,
>     prefix="/fixed_inputs",
>     tags=["Core", "Fixed Inputs"],
>     #dependencies=[Depends(authorization.authorize)],
>     dependencies=[Depends(get_authorization_function)],
> )
> ```
> 
> Where `get_authorization_function()` calls the `get_authorization()` at runtime instead of import-time. Otherwise the `authorization` object contained the OPA instance from oauth2-lib instead of our custom one.

Thanks for pointing that out! I didn’t notice this issue because I've been using the AuthZ from oauth2_lib without making any changes to it. I'll make sure to fix it.

Reply:
> In orchestrator-core I had to change the api code like below;
> 
> ```python
> #from orchestrator.security import get_authorization
> from orchestrator.security import get_authorization_function
> 
> # authorization = get_authorization()
> 
> api_router = APIRouter()
> api_router.include_router(
>     fixed_input.router,
>     prefix="/fixed_inputs",
>     tags=["Core", "Fixed Inputs"],
>     #dependencies=[Depends(authorization.authorize)],
>     dependencies=[Depends(get_authorization_function)],
> )
> ```
> 
> Where `get_authorization_function()` calls the `get_authorization()` at runtime instead of import-time. Otherwise the `authorization` object contained the OPA instance from oauth2-lib instead of our custom one.

But the get_authorization() function uses your settings to decide which authorization instance to load. I think it loads the default one unless you change your settings to use something other than oauth2_lib.fastapi.opa_instance. Am I missing something here?

Reply:
> I tried many ways to see if I could avoid using Depends from FastAPI. 

I agree that from an oauth2-lib point of view, the Auth code should ideally be framework agnostic whereas it is now `fastapi` specific. But we have little usecase for that at the moment, maybe in the future we'll need that if some of us want to switch to `fasterapi` or something :)

> But the get_authorization() function uses your settings to decide which authorization instance to load. I think it loads the default one unless you change your settings to use something other than oauth2_lib.fastapi.opa_instance. Am I missing something here?

I see what you're saying, I agree the settings object should have been initialised with our own instance string value. Not sure what happened, maybe I messed up somewhere..

Reply:
> I agree that from an oauth2-lib point of view, the Auth code should ideally be framework agnostic whereas it is now `fastapi` specific. But we have little usecase for that at the moment, maybe in the future we'll need that if some of us want to switch to `fasterapi` or something :)

I also looked into the oauthlib package (though I'm not completely sure about the name) and noticed that it has specific code for Django, Flask, and FastAPI. So having code for FastAPI isn't really a bad thing. But you do have a good point, and maybe in the future, we could think about making the code more general.

Reply:
I applied the suggestion and I think it made the code cleaner